src/client/datascience/jupyter/kernels/kernelProvider.ts

```diff
@@ -42,8 +42,21 @@ export class KernelProvider implements IKernelProvider {
     }
     public getOrCreate(uri: Uri, options: KernelOptions): IKernel | undefined {
         const existingKernelInfo = this.kernelsByUri.get(uri.toString());
-        if (existingKernelInfo && fastDeepEqual(existingKernelInfo.options.metadata, options.metadata)) {
-            return existingKernelInfo.kernel;
+        if (existingKernelInfo) {
+            if (
+                existingKernelInfo.options.metadata.kind === 'startUsingKernelSpec' &&
+                options.metadata.kind === 'startUsingKernelSpec'
+            ) {
+                // When using a specific kernelspec, just compare the actual kernel specs
+                if (fastDeepEqual(existingKernelInfo.options.metadata.kernelSpec, options.metadata.kernelSpec)) {
```

Comment:
So what was causing the issue here.

I.e. under what condition was the old code failing?
I'm trying to understand what was different in other properties of the `options.metadata` object that was different and cause the bug.

Reply:
Oh noticed instructions here https://github.com/microsoft/vscode-jupyter/issues/4109#issuecomment-742043965

Reply:
Cant we use `getKernelConnectionId` when comparing the kernel connection metadata instead of a deep compare

Reply:
i.e. `getKernelConnectionId(existingKernelInfo.options.metadata) === getKernelConnectionId(options.metadata`

This way it will handle comparing of connections that deal with live kernels (we only compare the `metata.kernelModel` property

Reply:
This PR might also be required/handy https://github.com/microsoft/vscode-jupyter/pull/4192/files